From 4c13da7fd6fb9fa9e0e6ce08897f9fa8ab164610 Mon Sep 17 00:00:00 2001
From: Esther Z <Esther.Z@linux.alibaba.com>
Date: Sun, 1 Sep 2024 00:30:01 +0800
Subject: [PATCH 198/400] dts: th1520: add LA regulator support.

scan camera, rgb camera, ir camera, lcd0\1, lcd bias, tp0\1, usb bus,
audio adc\dac are supported.

Signed-off-by: Esther Z <Esther.Z@linux.alibaba.com>
Signed-off-by: Han Gao <gaohan@iscas.ac.cn>
---
 .../boot/dts/thead/th1520-a-val-audio.dtsi    |  30 +--
 arch/riscv/boot/dts/thead/th1520-a-val.dtsi   | 191 +++++++++++-------
 2 files changed, 129 insertions(+), 92 deletions(-)

diff --git a/arch/riscv/boot/dts/thead/th1520-a-val-audio.dtsi b/arch/riscv/boot/dts/thead/th1520-a-val-audio.dtsi
index 5149b142416f..943218a3e9e6 100644
--- a/arch/riscv/boot/dts/thead/th1520-a-val-audio.dtsi
+++ b/arch/riscv/boot/dts/thead/th1520-a-val-audio.dtsi
@@ -259,9 +259,9 @@ es8156_audio_codec: es8156@8 {
 			compatible = "everest,es8156";
 			reg = <0x08>;
 			sound-name-prefix = "ES8156";
-			AVDD-supply = <&reg_aud_3v3>;
-			DVDD-supply = <&soc_dvdd18_aon_reg>;
-			PVDD-supply = <&soc_dvdd18_aon_reg>;
+		    	AVDD-supply = <&reg_aud_dac_3v3>;
+		   	DVDD-supply = <&soc_dvdd18_aon_reg>;
+		    	PVDD-supply = <&soc_dvdd18_aon_reg>;
 			mclk-sclk-ratio = <4>;
 			status = "disabled";
 		};
@@ -274,9 +274,9 @@ es7210_audio_codec_adc0: es7210@40 {
 			channels-max = <2>;
 			mclk-sclk-ratio = <4>;
 			sound-name-prefix = "ES7210_ADC0";
-			MVDD-supply = <&reg_aud_3v3>;
-			AVDD-supply = <&reg_aud_3v3>;
-			DVDD-supply = <&soc_dvdd18_aon_reg>;
+        		MVDD-supply = <&reg_aud_adc_3v3>;
+			AVDD-supply = <&reg_aud_adc_3v3>;
+		    	DVDD-supply = <&soc_dvdd18_aon_reg>;
 			PVDD-supply = <&soc_dvdd18_aon_reg>;
 			status = "disabled";
 		};
@@ -289,9 +289,9 @@ es7210_audio_codec_adc1: es7210@41 {
 			channels-max = <2>;
 			mclk-sclk-ratio = <4>;
 			sound-name-prefix = "ES7210_ADC1";
-			MVDD-supply = <&reg_aud_3v3>;
-			AVDD-supply = <&reg_aud_3v3>;
-			DVDD-supply = <&soc_dvdd18_aon_reg>;
+        		MVDD-supply = <&reg_aud_adc_3v3>;
+			AVDD-supply = <&reg_aud_adc_3v3>;
+		    	DVDD-supply = <&soc_dvdd18_aon_reg>;
 			PVDD-supply = <&soc_dvdd18_aon_reg>;
 			status = "disabled";
 		};
@@ -303,9 +303,9 @@ es7210_adc2: es7210@42 {
 			work-mode = "ES7210_TDM_1LRCK_DSPB";
 			channels-max = <8>;
 			sound-name-prefix = "ES7210_ADC2";
-			MVDD-supply = <&reg_aud_3v3>;
-			AVDD-supply = <&reg_aud_3v3>;
-			DVDD-supply = <&soc_dvdd18_aon_reg>;
+        		MVDD-supply = <&reg_aud_adc_3v3>;
+			AVDD-supply = <&reg_aud_adc_3v3>;
+		    	DVDD-supply = <&soc_dvdd18_aon_reg>;
 			PVDD-supply = <&soc_dvdd18_aon_reg>;
 			status = "disabled";
 		};
@@ -317,9 +317,9 @@ es7210_adc3: es7210@43 {
 			work-mode = "ES7210_TDM_1LRCK_DSPB";
 			channels-max = <8>;
 			sound-name-prefix = "ES7210_ADC3";
-			MVDD-supply = <&reg_aud_3v3>;
-			AVDD-supply = <&reg_aud_3v3>;
-			DVDD-supply = <&soc_dvdd18_aon_reg>;
+        		MVDD-supply = <&reg_aud_adc_3v3>;
+			AVDD-supply = <&reg_aud_adc_3v3>;
+		    	DVDD-supply = <&soc_dvdd18_aon_reg>;
 			PVDD-supply = <&soc_dvdd18_aon_reg>;
 			status = "disabled";
 		};
diff --git a/arch/riscv/boot/dts/thead/th1520-a-val.dtsi b/arch/riscv/boot/dts/thead/th1520-a-val.dtsi
index fbafe2d6fb39..571f7434eda1 100644
--- a/arch/riscv/boot/dts/thead/th1520-a-val.dtsi
+++ b/arch/riscv/boot/dts/thead/th1520-a-val.dtsi
@@ -50,117 +50,88 @@ chosen {
 		stdout-path = "serial0:115200n8";
 	};
 
-	reg_tp0_pwr: regulator-tp0-pwr {
+	reg_lcd0_vdd18: regulator-lcd0-vdd18 {
 		compatible = "regulator-fixed";
-		regulator-name = "tp0-pwr";
-		regulator-min-microvolt = <2800000>;
-		regulator-max-microvolt = <2800000>;
-		// gpio = <&ioexp3 4 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
-		enable-active-high;
-		regulator-always-on;
-	};
-
-	reg_hub_5v: regulator-hub-5v {
-		compatible = "regulator-fixed";
-		regulator-name = "HUB_5V";
-		regulator-min-microvolt = <5000000>;
-		regulator-max-microvolt = <5000000>;
-		// gpio = <&ioexp3 3 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
+		regulator-name = "lcd0_en";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		gpio = <&pcal6408ahk_a 2 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
-		regulator-always-on;
 	};
 
-	reg_vcc5v_usb: regulator-vcc5v-usb {
+	reg_tp0_pwr: regulator-tp0-pwr {
 		compatible = "regulator-fixed";
-		regulator-name = "VCC5V_USB";
-		regulator-min-microvolt = <5000000>;
-		regulator-max-microvolt = <5000000>;
-		gpio = <&gpio1 22 GPIO_ACTIVE_HIGH>;
+		regulator-name = "tp0_pwr";
+		regulator-min-microvolt = <2800000>;
+		regulator-max-microvolt = <2800000>;
+		gpio = <&pcal6408ahk_a 3 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
 		regulator-always-on;
 	};
 
-	reg_vdd33_lcd0: regulator-vdd33-lcd0 {
+	reg_lcd0_5v7: regulator-lcd0-vspn57 {
 		compatible = "regulator-fixed";
-		regulator-name = "lcd0_vdd33";
-		regulator-min-microvolt = <3300000>;
-		regulator-max-microvolt = <3300000>;
-		// gpio = <&ioexp3 5 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
+		regulator-name = "lcd0_bias_en";
+		regulator-min-microvolt = <5700000>;
+		regulator-max-microvolt = <5700000>;
+		gpio = <&pcal6408ahk_a 4 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
 	};
 
-	reg_vdd18_lcd0: regulator-vdd18-lcd0 {
+	reg_lcd1_1v8: regulator-lcd1-vdd18 {
 		compatible = "regulator-fixed";
-		regulator-name = "lcd0_vdd18";
+		regulator-name = "lcd1_en";
 		regulator-min-microvolt = <1800000>;
 		regulator-max-microvolt = <1800000>;
-		// gpio = <&ioexp3 6 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
+		gpio = <&pcal6408ahk_a 5 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
 	};
 
-	reg_vdd_3v3: regulator-vdd-3v3 {
+	reg_tp1_pwr: regulator-tp1-pwr {
 		compatible = "regulator-fixed";
-		regulator-name = "vdd_3v3";
-		regulator-min-microvolt = <3300000>;
-		regulator-max-microvolt = <3300000>;
-		gpio = <&gpio1 24 GPIO_ACTIVE_HIGH>;
+		regulator-name = "tp1_pwr";
+		regulator-min-microvolt = <2800000>;
+		regulator-max-microvolt = <2800000>;
+		gpio = <&pcal6408ahk_a 6 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
-		regulator-boot-on;
 		regulator-always-on;
 	};
 
-	reg_vref_1v8: regulator-adc-verf {
+	reg_lcd1_5v7: regulator-lcd1-vspn57 {
 		compatible = "regulator-fixed";
-		regulator-name = "vref-1v8";
-		regulator-min-microvolt = <1800000>;
-		regulator-max-microvolt = <1800000>;
-		regulator-alaways-on;
-		vin-supply = <&reg_vdd_3v3>;
+		regulator-name = "lcd1_bias_en";
+		regulator-min-microvolt = <5700000>;
+		regulator-max-microvolt = <5700000>;
+		gpio = <&pcal6408ahk_a 7 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
 	};
 
-	reg_aud_3v3: regulator-aud-3v3 {
+	reg_aud_adc_3v3: regulator-aud-adc-3v3 {
 		compatible = "regulator-fixed";
-		regulator-name = "aud_3v3";
+		regulator-name = "aud_adc_3v3_en";
 		regulator-min-microvolt = <3300000>;
 		regulator-max-microvolt = <3300000>;
+		gpio = <&pcal6408ahk_b 1 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
-		regulator-always-on;
-	};
-
-	reg_aud_1v8: regulator-aud-1v8 {
-		compatible = "regulator-fixed";
-		regulator-name = "aud_1v8";
-		regulator-min-microvolt = <1800000>;
-		regulator-max-microvolt = <1800000>;
-		enable-active-high;
-		regulator-always-on;
 	};
 
-	reg_cam0_dvdd12: regulator-cam0-dvdd12 {
+	reg_aud_dac_3v3: regulator-aud-dac-3v3 {
 		compatible = "regulator-fixed";
-		regulator-name = "dvdd12_cam0";
-		regulator-min-microvolt = <1200000>;
-		regulator-max-microvolt = <1200000>;
-		// gpio = <&ioexp1 0 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
-		enable-active-high;
-	};
-
-	reg_cam0_avdd28: regulator-cam0-avdd28 {
-		compatible = "regulator-fixed";
-		regulator-name = "avdd28_cam0";
-		regulator-min-microvolt = <2800000>;
-		regulator-max-microvolt = <2800000>;
-		// gpio = <&ioexp1 1 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
+		regulator-name = "aud_dac_3v3_en";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&pcal6408ahk_b 2 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
 	};
 
-	reg_cam0_dovdd18: regulator-cam0-dovdd18 {
+	reg_vcc5v_usb: regulator-vcc5v-usb {
 		compatible = "regulator-fixed";
-		regulator-name = "dovdd18_cam0";
-		regulator-min-microvolt = <2800000>;
-		regulator-max-microvolt = <2800000>;
-		// gpio = <&ioexp1 2 GPIO_ACTIVE_HIGH>;	to be FIXED !!!
+		regulator-name = "VCC5V_USB";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&gpio0 27 GPIO_ACTIVE_HIGH>;
 		enable-active-high;
+		regulator-always-on;
 	};
 
 	mbox_910t_client2: mbox_910t_client2 {
@@ -293,7 +264,7 @@ touch@5d {
 			interrupts = <8 IRQ_TYPE_LEVEL_LOW>;
 			irq-gpios = <&gpio1 8 0>;
 			reset-gpios = <&gpio1 7 0>;
-			// AVDD28-supply = <&reg_tp_pwr_en>;	to be FIXED !!!
+			AVDD28-supply = <&reg_tp0_pwr>;
 			status = "disbale";
 	};
 };
@@ -313,7 +284,7 @@ touch1@5d {
 		interrupts = <12 IRQ_TYPE_LEVEL_LOW>;
 		irq-gpios = <&gpio1 12 0>;
 		reset-gpios = <&gpio1 11 0>;
-		// AVDD28-supply = <&reg_tp1_pwr_en>;	to be FIXED !!!
+		AVDD28-supply = <&reg_tp1_pwr>;
 		status = "disbale";
 	};
 };
@@ -864,7 +835,6 @@ hub_2_0: hub@1 {
 		compatible = "usb2109,2817";
 		reg = <1>;
 		peer-hub = <&hub_3_0>;
-		vdd-supply = <&reg_hub_5v>;
 		vbus-supply = <&reg_vcc5v_usb>;
 	};
 
@@ -1351,7 +1321,7 @@ soc_vdd33_emmc_reg: soc_vdd33_emmc {
 	soc_vdd18_emmc_reg: soc_vdd18_emmc {
 		compatible = "thead,th1520-aon-pmic";
 		regulator-name = "soc_dvdd18_emmc";
-		regulator-type = "gpio";
+		regulator-type = "common";
 		regulator-boot-on;
 		regulator-always-on;
 	};
@@ -1375,10 +1345,77 @@ soc_dvdd12_scan_reg: soc_dvdd12_scan {
 	soc_avdd28_scan_en_reg: soc_avdd28_scan_en {
 		compatible = "thead,th1520-aon-pmic";
 		regulator-name = "soc_avdd28_scan_en";
+		regulator-type = "gpio";
+	};
+
+    	soc_vext_2v8_reg: soc_vext_2v8 {
+	    	compatible = "thead,th1520-aon-pmic";
+	    	regulator-name = "soc_vext_2v8";
+	    	regulator-type = "common";
+	    	regulator-boot-on;
+		regulator-always-on;
+	};
+
+	soc_avdd28_rgb_reg: soc_avdd28_rgb {
+		compatible = "thead,th1520-aon-pmic";
+		regulator-name = "soc_avdd28_rgb";
 		regulator-type = "common";
-		regulator-min-microvolt = <900000>;
+		regulator-min-microvolt = <2200000>;
+		regulator-max-microvolt = <3475000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	soc_dovdd18_rgb_reg: soc_dovdd18_rgb {
+		compatible = "thead,th1520-aon-pmic";
+		regulator-name = "soc_dovdd18_rgb";
+		regulator-type = "common";
+		regulator-min-microvolt = <1200000>;
+		regulator-max-microvolt = <3600000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	soc_dvdd12_rgb_reg: soc_dvdd12_rgb {
+		compatible = "thead,th1520-aon-pmic";
+		regulator-name = "soc_dvdd12_rgb";
+		regulator-type = "common";
+		regulator-min-microvolt = <400000>;
+		regulator-max-microvolt = <1675000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	soc_avdd25_ir_reg: soc_avdd25_ir {
+		compatible = "thead,th1520-aon-pmic";
+		regulator-name = "soc_avdd25_ir";
+		regulator-type = "common";
+		regulator-min-microvolt = <2200000>;
+		regulator-max-microvolt = <3475000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	soc_dovdd18_ir_reg: soc_dovdd18_ir {
+		compatible = "thead,th1520-aon-pmic";
+		regulator-name = "soc_dovdd18_ir";
+		regulator-type = "common";
+		regulator-min-microvolt = <1200000>;
 		regulator-max-microvolt = <3600000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	soc_dvdd12_ir_reg: soc_dvdd12_ir {
+		compatible = "thead,th1520-aon-pmic";
+		regulator-name = "soc_dvdd12_ir";
+		regulator-type = "common";
+		regulator-min-microvolt = <400000>;
+		regulator-max-microvolt = <1675000>;
+		regulator-boot-on;
+		regulator-always-on;
 	};
+
 };
 
 &dmac0 {
-- 
2.43.0

